partner_validate_vat/models/res_partner.py

```diff
@@ -0,0 +1,29 @@
+# coding: utf-8
+# Copyright 2017 Vauxoo (https://www.vauxoo.com) <info@vauxoo.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from openerp import models, api, fields, _
+from openerp.exceptions import ValidationError
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    vat = fields.Char(copy=False)
+
+    @api.constrains('vat', 'country_id')
+    def _check_vat_uniqueness(self):
+        """ Check that the vat is unique in the level
+            where the partner in the tree
+        """
+
+        user_company = self.env.user.company_id
+
+        for partner in self:
+            current_vat = partner.commercial_partner_id.vat
+            if (user_company.country_id == partner.country_id and
+                current_vat and partner.search([
+                    ('commercial_partner_id', '!=',
+                     partner.commercial_partner_id.id)], limit=1)):
```

Comment:
Verify your tests because they don't detect this error and this is the main focus of this module.